### PR TITLE
Fix #1890: Add tests for FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProvider

### LINKED ISSUE #1890: Thoroughly test FractionInputIsEquivalentToAndInSimplestFormRuleClassifier
Since the FractionInputIsEquivalentToAndInSimplestFormRuleClassifier was copied from Oppia web, it didn't have corresponding tests added to verify correctness. These should be ported over from Oppia web as well to ensure that each classifier has working functionality.

Useful reference - https://github.com/oppia/oppia-android/tree/develop/domain/src/test/java/org/oppia/android/domain/classify/rules

### PR DESCRIPTION
**Pull request by jcqli**
## Explanation
Fix #1890: Add tests for FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProvider
Also fixes part of #210
Tests that answer provided by learner is the simplest form and equivalent to the input given by creator -- even if input is not in simplest form, entered answer must still be in simplest form. Answers that are equal to input but _not_ simplest should not match. Checks for fractions, mixed numbers, whole numbers, and one negative case for each of these types of numbers.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      )
+    }
+
+    Truth.assertThat(exception)
+      .hasMessageThat()
+      .contains("Expected classifier inputs to contain parameter with name 'f' but had: [y]")
+  }
+
+  private fun setUpTestApplicationComponent() {
+    DaggerFractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest_TestApplicationComponent
```
Comment (jcqli) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: This line is longer than 100 chars, even without indentation...

**Comment by jcqli** [comment]
@vinitamurthi @BenHenning PTAL :)

**Comment by anandwana001** [comment]
@jcqli You can setup ktlint on your local machine, it helps to prevent the failure here. The pre-push is in the pipeline. Also, you can setup a macro which helps in re-formatting and removing the unused import. 
Check our wiki guide here - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,283 @@
+package org.oppia.android.domain.classify.rules.fractioninput
+
+import android.app.Application
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth
```
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: `import com.google.common.truth.Truth.assertThat`
Adding a nit comment before Ben reviews the PR.

Comment (jcqli) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Done

**Review by jcqli — COMMENTED** [review]
> @jcqli You can setup ktlint on your local machine, it helps to prevent the failure here. The pre-push is in the pipeline. Also, you can setup a macro which helps in re-formatting and removing the unused import.
> Check our wiki guide here - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide

Very helpful, thank you!

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 1 above)
Comment (jcqli) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Disabled ktlint for this line :)

**Comment by jcqli** [comment]
@anandwana001 I can't set you as the assignee right now but PTAL!

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    val inputs = mapOf("f" to FRACTION_VALUE_TEST_1_OVER_2)
+    val answer = FRACTION_VALUE_TEST_4_OVER_8
+
+    val matches =
+      inputIsEquivalentToAndInSimplestFormRuleClassifier.matches(answer = answer, inputs = inputs)
+
+    assertThat(matches).isFalse()
+  }
+
+  // Even if creator does not input simplest form, learner's answer must be in simplest form
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Suggest moving this down to be before the assert line to justify it matches. Also, maybe add 'answer must still be in simplest form'

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    val answer = MIXED_NUMBER_VALUE_TEST_16_AND_1_OVER_2
+
+    val matches =
+      inputIsEquivalentToAndInSimplestFormRuleClassifier.matches(answer = answer, inputs = inputs)
+
+    assertThat(matches).isFalse()
+  }
+
+  @Test
+  fun testEquivalentAndSimplest_missingInputF_throwsException() {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:240]: Suggest also adding tests for the wrong type of input/wrong type of answer (see other classifier tests for a reference).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jcqli, this looks really good! Just had two comments.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+      operation()
+      fail("Expected to encounter exception of $type")
+    } catch (t: Throwable) {
+      if (type.isInstance(t)) {
+        return type.cast(t)
+      }
+      // Unexpected exception; throw it.
+      throw t
+    }
+  }
```
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:307]: If we are seeing this being used often, should we keep this function in the testing module? @BenHenning @anandwana001 thoughts?

Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:307]: Yes, I guess, I will create a good first issue for this.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 3 above)
Comment (jcqli) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 4 above)
Comment (jcqli) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:240]: Done

**Comment by jcqli** [comment]
@BenHenning PTAL!

**Review by BenHenning — APPROVED** [review]
Thanks @jcqli! This looks great!

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 5 above)
Comment (BenHenning) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:307]: Probably. :) Thanks @anandwana001. Can you reference that starter issue here once created?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Actually, there's one more thing needed here that I missed the first time around. @jcqli please add this to the Bazel BUILD file for the domain module (see the other similar tests for an example).

File: domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt (hunk 5 above)
Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputIsEquivalentToAndInSimplestFormRuleClassifierProviderTest.kt:307]: https://github.com/oppia/oppia-android/issues/2234

**Comment by jcqli** [comment]
@BenHenning added to build, PTAL!

File: domain/BUILD.bazel [hunk 6]
```diff
@@ -102,6 +102,96 @@ domain_test(
     deps = TEST_DEPS,
 )
 
+domain_test(
```
Comment (BenHenning) [inline domain/BUILD.bazel]: FYI that #2240 was just submitted. You'll probably get merge conflicts from this--can you update your changes with the latest develop?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jcqli! Just one comment--once that's resolved this will LGTM!

File: domain/BUILD.bazel (hunk 6 above)
Comment (jcqli) [reply domain/BUILD.bazel]: Done

**Comment by jcqli** [comment]
@BenHenning Missed this comment earlier! Build file updated, PTAL

**Review by BenHenning — APPROVED** [review]
Thanks @jcqli this LGTM!

**Comment by BenHenning** [comment]
Going ahead and merging this. @anandwana001 & @vinitamurthi if you have any follow-up comments, please go ahead and add them & we can address them in a follow-up PR.

**Pull request merged**